doc/guides/networking/net-stack-architecture.rst

```diff
@@ -0,0 +1,137 @@
+
+* **Network Application.** The network application uses
+  :ref:`BSD socket API <bsd_sockets_interface>` to create a network connection,
+  send or receive data, and close a connection. The application can also use
+  the :ref:`network management API <net_mgmt_interface>` to set network
+  related parameters such as taking network interface down, starting a scan
+  (when applicable), setting IP address to a network interface, etc.
+
+* **Network Protocols.** This provides implementations for
+  various protocols such as IPv6, IPv4, UDP, TCP, ICMPv4 and ICMPv6.
+  See :ref:`application protocols chapter <net_protocols>` for information
+  about CoAP, LWM2M and MQTT.
+
+* **Network Interface Abstraction.** This provides functionality
+  that is common in all the network interfaces, such as setting network
+  interface down, etc. There can be multiple network interfaces in the system.
+  See :ref:`network interface overview <net_if_interface>` for more details.
+
+* **L2 Network Technologies.** This provides common API for sending and
+  receiving data to and from an actual network device.
+  See :ref:`L2 overview <net_l2_interface>` for more details.
+  These network technologies include :ref:`Ethernet <ethernet_interface>`,
+  :ref:`IEEE 802.15.4 <ieee802154_interface>`, Bluetooth, CANBUS, etc.
+  Some of these technologies support IPv6 header compression (6Lo), which is
+  done in its own layer. For example ARP for IPv4 is done by the
+  :ref:`Ethernet component <ethernet_interface>`.
+
+* **Network Device Driver.** The actual low-level device driver handles the
+  physical sending or receiving of a network packet.
+
+
+Network data flow
+*****************
+
+The application typically consists of one or more threads that execute
+the application logic. When using the network connectivity APIs, following
+things will happen.
+
+.. figure:: zephyr_netstack_overview-rx_sequence.svg
+    :align: left
+    :alt: Network RX data flow
+    :figclass: align-center
+
+    Network RX data flow
+
+*Data receiving (RX):*
+
+1. A network data packet is received by a device driver.
+
+2. The device driver allocates enough network buffers to store the received
+   data. The network packet is placed to proper RX queue (implemented by
+   k_fifo). By default there is only one receive queue in the system,
+   but it is possible to have max. 8 receive queues. These queues will
+   process incoming packets with different priority.
+   See :ref:`traffic-class-support` for more details. The receive queues also
+   acts as a way to separate the data processing pipeline (bottom-half) as
+   the device driver is running in interrupt context and it must do its
+   processing as fast as possible.
+
+3. The network packet is then passed to correct L2 driver. The L2 driver can
+   check if the packet is proper and modify it if needed.
+
+4. The packet goes through network interface. There network statistics are
```

Comment:
There multiple tweaks I could suggest to this doc, but leave for later. But can't pass beyond this one. Can we please say "The packet is processed by a network interface." Because "goes thru" is ambiguous.

Reply:
Now would be a good time to suggest any tweaks you have, we can then discuss about them here. I can certainly change the wording about network interface.